src/track/beatutils.cpp

```diff
@@ -5,44 +5,350 @@
+        // tolerence we assume the region has a constant tempo and make a fixed tempo grid
+        if (leftRighDiff < kMaxDiffSameBpm) {
+            auto splittedAtTempoChange = makeQVector(
+                    rawBeats.begin() + beatStart, rawBeats.begin() + beatEnd + 1);
+            double bpm = calculateBpm(splittedAtTempoChange, sampleRate, minBpm, maxBpm);
+            ironedBeats = calculateFixedTempoGrid(splittedAtTempoChange, sampleRate, bpm);
+        }
+        // not const, make ironed grid of longest sequence withing a 25ms phase error
+        else {
+            auto splittedAtTempoChange = makeQVector(
+                    rawBeats.begin() + beatStart, rawBeats.begin() + beatEnd + 1);
+            ironedBeats = calculateIronedGrid(splittedAtTempoChange, sampleRate);
+        }
+        if (correctedBeats.size() > 0) {
+            if (correctedBeats.last() == ironedBeats.first()) {
+                correctedBeats.pop_back();
+            }
+        }
+        correctedBeats << ironedBeats;
+    }
+    return correctedBeats;
+}
+
+QVector<double> BeatUtils::calculateFixedTempoGrid(
+        const QVector<double>& rawbeats, int sampleRate, const double localBpm) {
+    if (rawbeats.size() < kBeatsToCountTempo) {
+        return rawbeats;
+    }
+    // Length of a beat at localBpm in mono samples.
+    const double beat_length = (60.0 * sampleRate) / localBpm;
+    QMap<double, int> tempoFrequency; 
+    QList<double> tempoList = computeWindowedBpmsAndFrequencyHistogram(
+            rawbeats, kBeatsToCountTempo, 1, sampleRate, &tempoFrequency);
+    int longestSequence = 0;
+    int longestSequenceEnd = 0;
+    int sequence = 0;
+    // we look for the longest sequence of beats that have the same tempo
+    // and consider them to have the right phase, might not be true if
+    // a tshack was detect instead of a boom - but there is nothing we can do
+    for (int i = 0; i < tempoList.size() - 1; i+=1) {
+        if (tempoList[i] == tempoList[i+1]) {
+            sequence += 1;
+        } else {
+            sequence = 0;
+        }
+        if (sequence > longestSequence) {
+            longestSequence = sequence;
+            longestSequenceEnd = i;
+        }
+    }
+    int leftIndex = longestSequenceEnd;
+    double leftBeatOffset = rawbeats[leftIndex];
+    QVector<double> fixedBeats;
+    while (leftIndex > 0) {
+        leftBeatOffset -= beat_length;
+        fixedBeats << leftBeatOffset;
+        leftIndex -= 1;
+    }
+    std::reverse(fixedBeats.begin(), fixedBeats.end());
+    int rightIndex = longestSequenceEnd + 1;
+    double rightBeatOffset = rawbeats[rightIndex];
+    while (rightIndex < rawbeats.size()) {
+        rightBeatOffset += beat_length;
+        fixedBeats << rightBeatOffset;
+        rightIndex += 1;
+    }
+    return fixedBeats;
+}
+
+QVector<double> BeatUtils::calculateIronedGrid(
+        const QVector<double>& rawbeats, const int sampleRate) {
+    // Daniel's red ironing algorithm
```

Comment:
red? What is red about this?

Reply:
More like an internal joke, it was red on the graph on the spreadsheet @daschuer made it, and he left me a big todo to implement the red algorithm. Should remove or ok to keep?

Reply:
Nobody who reads this code in 10 years is going to know that. Please remove it.